### ISSUE TITLE
[ban-types] can't ban types with type parameters

### ISSUE THREAD

**Issue by binoche9** [typescript-eslint/typescript-eslint#2657]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/ban-types": [
      "error",
      {
        "Foo<Bar>": "bad"
      }
    ]
  }
}
```

```TS
type Foo<T> = { x: T }
type X = Foo<Bar>
```

**Expected Result**

ESLint detects a lint error on line 2 because I used `Foo<Bar>`.

**Actual Result**

No error was reported.

**Additional Info**

I first looked into this because [the docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md#options) suggested that specifying a type parameter should be possible. However, looking at [the implementation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/ban-types.ts#L226-L228) the it's clear that rule doesn't look at the type parameter information (it passes in the `typeName` node of the `TSTypeReference` to `checkBannedTypes()`). I found [the PR](https://github.com/typescript-eslint/typescript-eslint/commit/1c0ce9b6110c1d93cd64a3465a7f18bc2f7c034b#diff-b1774bf749955baa6c74936f14b7a306R34) where the documentation was added and it seems the functionality in the doc was never implemented.

At the least I think the docs could be clarified, but it would be nice if we can ban a type from taking a certain parameter (in my case, `any`).

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.40` |
| `@typescript-eslint/parser`        | `4.40` |
| `TypeScript`                       | `4.03` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `14.5.0` |

**Comment by bradzacher** [comment]
> in my case, any

Why not just use our `no-explicit-any` rule?

**Comment by bradzacher** [comment]
I'm unable to repro this against master.
![image](https://user-images.githubusercontent.com/7462525/95599869-81481b00-0a06-11eb-8ab8-08bc447413e7.png)

Please provide an example which exactly repros the issue.

**Comment by binoche9** [comment]
Unfortunately, I can't use `no-explicit-any` because my existing codebase has too many explicit `any`s. I just want to restrict using it as a type parameter to a specific type.

My expectation from reading the docs is that the test should fail (because the code uses `Foo<Bar>` which is banned in the config) but the test passes. 

Honestly from my reading of the code, this looks like a case where the doc just states something that was never intended/implemented.

**Comment by bradzacher** [comment]
yeah i wasn't paying attention sorry. it was supposed to error ofc, so I did repro it.
looking at too many things at once.

---

I don't think there was ever an explicit test added for it, so I'm not sure if it ever worked, or if it was broken by some change. 🤷‍♂️ hard to say.

Regardless it's not working now. Happy to accept a PR to update the docs or to make it work - either or!
